🐋👌Add VC Redistributable to base images for python package compatibility

jsnel opened the pull request:
Update the Dockerfiles to bundle the latest `vc_redist.x64.exe` into the images.

This is needed for certain python packages which implicitly assume these (redistributable) binaries are present on the system (see #1 for a reproducible example with `matplotlib`). 

Closes #1

jsnel committed:
🐋👌 Add commands to install vc_redist.x64 into images during build

Downloads, installs and cleans up vc_redist.x64.exe.

The vc_redist.x64.exe is downloaded using the recommended URL provided by MS documentation.
Ref: "Microsoft Visual C++ Redistributable latest supported downloads"
The link "https://aka.ms/vs/17/release/vc_redist.x64.exe" is described as "permalink for latest supported x64 version."

jsnel committed:
📚🐋 Add metadata to Dockerfiles

Add OCI labels for metadata as per instructions:
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#labelling-container-images

jsnel committed:
📚 Update README

Added note to the readme about adding the latest (2015-2022) Visual Studio Redistributable.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
+LABEL org.opencontainers.image.url="https://github.com/s-weigand/uv-windows-docker"
+LABEL org.opencontainers.image.documentation="https://github.com/s-weigand/uv-windows-docker/blob/main/README.md"
+LABEL org.opencontainers.image.source=https://github.com/s-weigand/uv-windows-docker
+LABEL org.opencontainers.image.licenses="Apache-2.0"
```
Not sure we can use the `licenses` label.
While the code in this repo to build the images is `Apache-2.0`, it might not be the case for the resulting image, depending on if you consider the base image as part of the image or not, as well as if the tool (`uv` and `vcredist`) licenses should also be included in the label.
I looked at the official python windows images and [they don't use any labels](https://github.com/docker-library/python/blob/2589112d847139587507c8071a80b07134a4b372/3.12/windows/windowsservercore-ltsc2022/Dockerfile).
So let's just remove it for now.

```suggestion
```

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
+# OCI labels for metadata
+LABEL org.opencontainers.image.title="Windows Server LTSC 2022 Image pre-installed with uv and VC Redist"
+LABEL org.opencontainers.image.description="A Windows Server image with the uv python package and project manager from astral  installed, Python version 3.9-3.13 pre-installed, and including the Visual C++ Redistributable."
+LABEL org.opencontainers.image.version="0.0.2"
```
The version should be the same as the `uv` (and image tag) version
```suggestion
LABEL org.opencontainers.image.version="${UV_VERSION}"
```

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
+LABEL org.opencontainers.image.title="Windows Server LTSC 2022 Image pre-installed with uv and VC Redist"
+LABEL org.opencontainers.image.description="A Windows Server image with the uv python package and project manager from astral  installed, Python version 3.9-3.13 pre-installed, and including the Visual C++ Redistributable."
+LABEL org.opencontainers.image.version="0.0.2"
+# LABEL org.opencontainers.image.authors="Sebastian Weigand <s.weigand.phy@gmail.com>"
```
Let's remove this, if anyone find an issue they can use the GH issue tracker (don't want mails directly and backport them to the issue tracker for transparency 😅).

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
 
+RUN $vcRedistUrl='https://aka.ms/vs/17/release/vc_redist.x64.exe'; `
+    $installerPath='C:\vc_redist.x64.exe'; `
+    Invoke-WebRequest -Uri $vcRedistUrl -OutFile $installerPath -UseBasicParsing; `
```
Let's use `WebClient.DownloadFile` since it is a lot faster for big files.
```suggestion
    (New-Object System.Net.WebClient).DownloadFile($vcRedistUrl, $installerPath); `
```

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.servercore:
```diff
@@ -2,9 +2,24 @@
 
```
Same as for the other `Dockerfile`.

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on README.md:
```diff
@@ -1,2 +1,10 @@
+
+Additionally the latest (2015-2022) Visual Studio Redistributable has been added to the images.
+
+The repo builds 2 types of images: `server` and `servercore` with as basis respectively:
```
I think this makes it a bit more clear.

```suggestion
The repo builds 2 images: `server` and `servercore` with the respective base images:
```

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

Slightly reworded again in commit b356ba0c1e1903951e6863695c4396ed28eef08a

s-weigand commented on README.md:
```diff
@@ -1,2 +1,10 @@
+The repo builds 2 types of images: `server` and `servercore` with as basis respectively:
+
+- mcr.microsoft.com/windows/server:ltsc2022
+- mcr.microsoft.com/windows/servercore:ltsc2022
```
Minor readability improvement.
```suggestion
- `mcr.microsoft.com/windows/server:ltsc2022`
- `mcr.microsoft.com/windows/servercore:ltsc2022`
```

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
 
 FROM mcr.microsoft.com/windows/server:ltsc2022
 
+# OCI labels for metadata
```
We should move all `ARG`s right below the `FROM` image so they can be used it across the image without later reordering things. And the labels below the VCRedist installation for better FS layer caching.

jsnel replied:
Reviewer suggestion adopted and incorporated in commit 087e34076735b42bedc516abebf11506294b2b42.

s-weigand commented on docker/Dockerfile.server:
```diff
@@ -2,9 +2,24 @@
 
 SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
 
+RUN $vcRedistUrl='https://aka.ms/vs/17/release/vc_redist.x64.exe'; `
```
For layer caching this should be moved below enabling the long path support.

jsnel replied:
This was done in b22b8af0ce78533be6a3af08ea43e91eda05e9a4.

s-weigand reviewed (changes requested):
Thanks a lot for the PR, I did just build the image and didn't use it yet, so the common need for VCRedist did slip my attention.
I just have some minor improvement requests.

s-weigand commented:
@all-contributors please add @jsnel  for bug code doc

jsnel committed:
👌 Incorporate reviewer's suggestions

Reviewed on: https://github.com/s-weigand/uv-windows-docker/pull/3

All review comments made sense and were incorporated verbatim.

jsnel committed:
📚👌Improve README readability

Single sentence stating its purpose.

jsnel committed:
👌Reviewer suggestion: Improved layer caching

"For layer caching installing redist should be moved below enabling the long path support."

s-weigand reviewed (approved):
Thanks again for the improvement, all looks good to me now 🚀
